cca.go

```diff
@@ -0,0 +1,122 @@
+// panic.
+//
+// Canonical Correlations will panic if the inputs x and y do not have the same
+// number of rows, and their columns should be interepretable as two different
+// sets of measurements (variables) on the same observations (each row representing
+// an observation).
+//
+// The sphered data are Xc * Sx^(-1/2) and Yc * Sy^(-1/2) respectively, and the
+// correlation matrix between the sphered data is the canonical correlation matrix,
+// Sx^(-1/2) * Sxy * Sy^(-1/2). In cases where S^(-1/2) is ambiguous, S^(-1/2) is
+// taken to be E * D^(-1/2) * E' where S can be eigendecomposed as S = E * D * E'.
+// Xc and Yc denote the centered input data matrices x and y (column means subtracted
+// from each column), Sx and Sy denote the sample covariance matrices within x and y
+// respectively, and Sxy denotes the covariance matrix between x and y.
+//
+// The first pair of left and right eigenvectors of the canonical correlation matrix
+// can be interpreted as directions into which the respective sphered data can be
+// projected such that the correlation between the two projections is maximised.
+// The second pair and onwards solve the same optimisation but under the constraint
+// that they are uncorrelated to previous projections.
+//
+// More details can be found at https://en.wikipedia.org/wiki/Canonical_correlation
+// or in Chapter 3 of
+// Koch, Inge. Analysis of multivariate and high-dimensional data.
+// Vol. 32. Cambridge University Press, 2013. ISBN: 9780521887939
+//
+// If one of the necessary singular value decompositions is not possible, an error
+// is returned as err and all other outputs are returned as nil.
+func CanonicalCorrelations(x, y mat64.Matrix, weights []float64) (corrs []float64, pvecs, qvecs *mat64.Dense, xsvd, ysvd *mat64.SVD, err error) {
+	n, _ := x.Dims()
+	yn, _ := y.Dims()
+	if n != yn {
+		panic("stat: unequal number of observations")
+	}
+	if weights != nil && len(weights) != n {
+		panic("stat: len(weights) != observations")
+	}
+
+	var ok bool
+
+	// Centre and factorize x and y.
+	xsvd, ok = svdFactorizeCentered(x, weights)
+	if !ok {
+		return nil, nil, nil, nil, nil, errors.New("stat: failed to factorize x")
+	}
+	ysvd, ok = svdFactorizeCentered(y, weights)
+	if !ok {
+		return nil, nil, nil, nil, nil, errors.New("stat: failed to factorize y")
+	}
+	var xu, xv, yu, yv mat64.Dense
+	xu.UFromSVD(xsvd)
+	xv.VFromSVD(xsvd)
+	yu.UFromSVD(ysvd)
+	yv.VFromSVD(ysvd)
+
+	// Calculate and factorise the canonical correlation matrix.
+	var ccor mat64.Dense
+	ccor.Product(&xv, xu.T(), &yu, yv.T())
+	var csvd mat64.SVD
+	if ok := csvd.Factorize(&ccor, svdKind(ccor.Dims())); !ok {
+		return nil, nil, nil, nil, nil, errors.New("stat: failed to factorize ccor")
+	}
+	var pv, qv mat64.Dense
+	pv.UFromSVD(&csvd)
+	qv.VFromSVD(&csvd)
+	corrs = csvd.Values(nil)
+
+	return corrs, &pv, &qv, xsvd, ysvd, nil
+}
+
+func svdKind(n, d int) matrix.SVDKind {
```

Comment:
If we wanted to export them they could go with SVD.Factorise but we probably don't want to export them. They could go in pca.go but that doesn't make too much difference really. In principle we could merge pca.go and cca.go into one file but what would it be called? A minor point really, I suppose.